### DISCUSSION TITLE
loss functions and optimizers as hyperparameters

### DISCUSSION THREAD

**Discussion by daihaozxn** [optuna/optuna#4664] [Q&A] [answered]
Excuse me, may I use many loss functions and optimizers under the pytorch framework as hyperparameters, and use optuna to perform searches? 
Where could I find examples of this?

**Comment by nzw0301** [comment]
I think https://optuna.readthedocs.io/en/stable/tutorial/20_recipes/002_multi_objective.html is such an example.

**Reply by daihaozxn** [reply]
> I think https://optuna.readthedocs.io/en/stable/tutorial/20_recipes/002_multi_objective.html is such an example.

Thank you for the answer.
After reading the tutorial, I'm sorry maybe I didn't express my requirements clearly at the beginning.
I hope to use the pytorch loss function such as torch.nn.MSELoss(), torch.nn.L1Loss(), torch.nn.SmoothL1Loss() as a hyperparameter, and find the optimal loss function among them. (The situation of the optimizer as a hyperparameter is similar, that is, the optimizer of pytorch such as torch.optim.Adam, torch.optim.SGD, torch.optim.RMSprop is used as a hyperparameter to optimize in them.)
But the return value type of function trial.suggest_categorical is one of the five types None, int, float, str, bool. None of them are suitable for the pytorch loss function or optimizer.

**Reply by nzw0301** [reply, marked as answer]
Thank you for your clarification. In this case, we can do that by using `dict` as follows:


```python
losses = {
    "MSELoss": torch.nn.MSELoss(),
    "L1Loss": torch.nn.L1Loss(),
    "SmoothL1Loss": torch.nn.SmoothL1Loss(),
}

loss_key = suggest_categorical("loss", list(losses.keys()))

loss = losses[loss_key]
loss(pred, labels)
```

**Reply by daihaozxn** [reply]
> Thank you for your clarification. In this case, we can do that by using `dict` as follows:
> 
> ```python
> losses = {
>     "MSELoss": torch.nn.MSELoss(),
>     "L1Loss": torch.nn.L1Loss(),
>     "SmoothL1Loss": torch.nn.SmoothL1Loss(),
> }
> 
> loss_key = suggest_categorical("loss", list(losses.keys()))
> 
> loss = losses[loss_key]
> loss(pred, labels)
> ```

It can work now. Thank you very much.
